core/watcher/eventsource/key.go

```diff
@@ -42,6 +42,8 @@ func (w *KeyWatcher) Changes() watcher.NotifyChannel {
 }
 
 func (w *KeyWatcher) loop() error {
+	defer close(w.out)
```

Comment:
Can we have a generic test that ensures that a watcher closes its channel on kill?

```go
type ChangeWatcher[T any] interface {
	worker.Worker
	Changes() <-chan T
}

func EnsureCloseOnCleanKill[T any](c *gc.C, w ChangeWatcher[T]) {
	worker.CleanKill(c, w)
	_, ok := w.Changes()
	c.Assert(ok, jc.IsFalse)
}

func EnsureCloseOnDirtyKill[T any](c *gc.C, w ChangeWatcher[T]) {
	worker.DirtyKill(c, w)
	_, ok := w.Changes()
	c.Assert(ok, jc.IsFalse)
}
```

Reply:
As much as I love this proposal, and the time I spent trying to make it work, it's not going to be possible. 
The main reason is that our watchers' `Changes()` method return alias types ,`watcher.NotifyChannel` for instance (which are aliased to a channel), and therefore we cannot use the type parameter. If I use `watcher.NotifyChannel` as type param, then I have to change a bit your proposal and it still won't work because go doesn't allow for casting any type to a channel type:
```go
type ChangeWatcher[T any] interface {
    worker.Worker
    Changes() T
}

func EnsureCloseOnCleanKill[T any](c *gc.C, w ChangeWatcher[T]) {
    workertest.DirtyKill(c, w)
    switch t := any(w.Changes()).(type) {
    case any:
        new, ok := (t).(<-chan struct{})
        c.Assert(ok, jc.IsTrue) // THIS IS FALSE (cannot cast)
        _, ok = <-new
        c.Assert(ok, jc.IsFalse)
    case <-chan T:
        _, ok := <-t
        c.Assert(ok, jc.IsFalse)
    }
}
```

Reply:
so I'm leaving the small checks inside the separated tests without adding any generic helpers function